c error when the method accept the reference but it is passed a lvalue

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->
**V doctor:**
```
OS: linux, Linux Mint 20.2
Processor: 4 cpus, 64bit, little endian, Intel(R) Core(TM) i5-6300U CPU @ 2.40GHz
CC version: cc (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0

getwd: /home/vincent/Github/vlang/v
vmodules: /home/vincent/.vmodules
vroot: /home/vincent/Github/vlang/v
vexe: /home/vincent/Github/vlang/v/v
vexe mtime: 2022-05-04 10:39:11
is vroot writable: true
is vmodules writable: true
V full version: V 0.2.4 3bd6455.e968e2b

Git version: git version 2.25.1
Git vroot status: weekly.2022.18-10-ge968e2be-dirty
.git/config present: true
thirdparty/tcc status: thirdparty-linux-amd64 eee03626

```

**What did you do?**
`v -g -o vdbg cmd/v && vdbg vlib/datatypes/rbtree_test.v`
```v
fn test_insert_into_rbtree() {
	mut rbtree := RBTree<int>{}
        // the method expected a &value
	assert rbtree.insert(12)
}

```

**What did you expect to see?**

a checker error that tell that the lvalue can not be passed by reference


**What did you see instead?**
```
==================
/tmp/v_1000/rbtree_test.11456963412675084342.tmp.c:12637: error: lvalue expected
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang

```

--- a/vlib/v/checker/fn.v
+++ b/vlib/v/checker/fn.v
@@ -897,6 +897,11 @@ pub fn (mut c Checker) fn_call(mut node ast.CallExpr, mut continue_check &bool) 
 			}
 			continue
 		}
+		if param.typ.is_ptr() && !param.is_mut && !call_arg.typ.is_real_pointer()
+			&& call_arg.expr.is_literal() && func.language == .v {
+			c.error('literal argument cannot be passed as reference parameter `${c.table.type_to_str(param.typ)}`',
+				call_arg.pos)
+		}
 		c.check_expected_call_arg(arg_typ, c.unwrap_generic(param.typ), node.language,
 			call_arg) or {
 			if param.typ.has_flag(.generic) {
@@ -1432,6 +1437,10 @@ pub fn (mut c Checker) method_call(mut node ast.CallExpr) ast.Type {
 					}
 				}
 				continue
+			}
+			if param.typ.is_ptr() && !arg.typ.is_real_pointer() && arg.expr.is_literal() {
+				c.error('literal argument cannot be passed as reference parameter `${c.table.type_to_str(param.typ)}`',
+					arg.pos)
 			}
 			c.check_expected_call_arg(got_arg_typ, exp_arg_typ, node.language, arg) or {
 				// str method, allow type with str method if fn arg is string

--- a/vlib/v/checker/tests/fn_ref_arg_mismatch_err.out
+++ b/vlib/v/checker/tests/fn_ref_arg_mismatch_err.out
@@ -1,0 +1,13 @@
+vlib/v/checker/tests/fn_ref_arg_mismatch_err.vv:15:10: error: literal argument cannot be passed as reference parameter `&T`
+   13 | fn main() {
+   14 |     foo := Foo<int>{}
+   15 |     foo.foo(12)
+      |             ~~
+   16 |
+   17 |     bar<int>(12)
+vlib/v/checker/tests/fn_ref_arg_mismatch_err.vv:17:11: error: literal argument cannot be passed as reference parameter `&T`
+   15 |     foo.foo(12)
+   16 |
+   17 |     bar<int>(12)
+      |              ~~
+   18 | }

--- a/vlib/v/checker/tests/fn_ref_arg_mismatch_err.vv
+++ b/vlib/v/checker/tests/fn_ref_arg_mismatch_err.vv
@@ -1,0 +1,18 @@
+module main
+
+struct Foo<T> { }
+
+fn (f &Foo<T>) foo(a &T) {
+	println(a)
+}
+
+fn bar<T>(a &T) {
+	println(a)
+}
+
+fn main() {
+	foo := Foo<int>{}
+	foo.foo(12)
+
+	bar<int>(12)
+}

